### PR TITLE
Show message when using unsupported by-name parameters

### LINKED ISSUE #7732: Error in the provide macro when using case class copy
While migrating specs onto ZIO 2, we've encountered the following errors:
```
[error] Error while emitting ~~~~1Spec.scala
[error] value x$11
[error] Error while emitting ~~~~2Spec.scala
[error] value x$304
[error] Error while emitting ~~~~3Spec.scala
[error] value x$2
```

After deleting every line until we found what the problem is, we discovered that invoking `copy` on a config class in the `provide` section caused the error:
```scala
.provide(
    ...
    ZLayer.succeed(config.copy(eodLagLimit = 10.days)),
    ...
```
causes an issue, but:
```scala
.provide(
    ...
    ZLayer.succeed(config),
    ...
```
doesn't.

### PR DESCRIPTION
**Pull request by andrzejressel**
It's not really fixable on ZIO side, but at least better error message can be shown.

Fixes #7732
/claim #7732

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -255,6 +255,10 @@ lazy val coreTests = crossProject(JSPlatform, JVMPlatform, NativePlatform)
   .dependsOn(testRunner)
   .settings(buildInfoSettings("zio"))
   .settings(publish / skip := true)
+  .settings(
```
Comment (adamgfraser) [inline build.sbt]: You can use the `typecheck` operator in ZIO Test for this. There is no need for a third party dependency.

**Comment by andrzejressel** [comment]
I forgot about copy one ... I'll add it

EDIT: I've tested copy and cannot reproduce it - my guess is that propably it was only the by-name thing

File: build.sbt (hunk 1 above)
Comment (andrzejressel) [reply build.sbt]: Done

File: core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala [hunk 2]
```diff
@@ -9,11 +9,22 @@ private[zio] trait LayerMacroUtils {
 
   type LayerExpr = c.Expr[ZLayer[_, _, _]]
 
+  private def verifyLayers(layers: Seq[c.Expr[ZLayer[_, _, _]]]): Unit =
+    for (layer <- layers) {
+      layer.tree match {
+        case apply @ Apply(_, Block(_, _) :: Nil) => {
+          c.abort(layer.tree.pos, "Cannot use by-name parameters")
```
Comment (jdegoes) [inline core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala]: Can you tweak this error message to explain:

1. What's going wrong, in detail (a user might not have ever heard of "by-name" parameters)
2. Why it's going wrong, e.g., what problem is introduced by by-name parameters that prevents their invocation from appearing as input to provide
3. What the user can do to fix the problem

Comment (andrzejressel) [reply core/shared/src/main/scala-2/zio/internal/macros/LayerMacroUtils.scala]: I've created following error message:
```
──── ZLAYER ERROR ────────────────────────────────────────────────────

 Due to bug in Scala 2 compiler invoking methods with by-name parameters in provide/provideSome method does not work
 def createLayerByName(i: Int, x: => MyLayer): zio.ULayer[MyLayer]

 Bug can be workarounded in following ways:

 1. Assign method output to temporary variable
    ZLayer.provide(createLayerByName(...))
    ↓↓↓
    val temp = createLayerByName(...)
    ZLayer.provide(temp)

 2. Replace by-name parameter with lambda:
    def createLayerByName(i: Int, x: => MyLayer): zio.ULayer[MyLayer]
    ↓↓↓
    def createLayerByName(i: Int, x: () => MyLayer): zio.ULayer[MyLayer]

──────────────────────────────────────────────────────────────────────
```

**Review by jdegoes — APPROVED** [review]

**Pull request merged**